### 프리플롭

In [1]:
from math import comb

# Calculate the probability of getting a pocket pair, connectors, suited pre-flop in Texas Hold'em

# Total number of ways to pick any two cards from the deck
total_two_card_combinations = comb(52, 2)

# One_pair
# For a specific rank, there are 4 cards, and you need 2 of them to make a pocket pair
ways_to_pick_pocket_pair_per_rank = comb(4, 2) * 13

# Connectors
# There are 12 possible connector pairs (A-K not considered here because it is not a straight connector in Texas Hold'em)
# Each pair can be in any of 4 suits, and each card of the pair can also be in any of 4 suits
ways_to_pick_connectors = 12 * 4 * 4
'''
2-3에서 A-K까지, 총 12개의 커넥터 쌍이 가능합니다.
각 카드가 4가지 무늬 중 하나를 가질 수 있으므로, 두 카드 각각에 대해 4가지 선택이 가능합니다.
따라서 12개의 커넥터 쌍 각각에 대해 무늬 조합이 4 (첫 번째 카드의 무늬) × 4 (두 번째 카드의 무늬) = 16가지 방법이 있습니다. 
이를 모든 12개의 커넥터 쌍에 적용하면, 12 × 4 × 4 = 192가지 조합이 나옵니다.
'''
# Suited cards
# For each suit, choose 2 out of 13 cards
ways_to_pick_suited = 4 * comb(13, 2)
'''
Suited는 같은 무늬의 두 카드를 의미합니다.
한 무늬에는 13장의 카드가 있으므로, 같은 무늬를 가진 두 카드를 선택하는 조합은 13개 중 2개를 선택하는 조합입니다.
각 무늬별로 78가지 조합이 가능합니다 
4 무늬가 있으므로 총 312가지 조합이 나옵니다
'''

# Suited connectors
# There are 12 possible suited connector pairs
# Each pair can be in any of 4 suits
ways_to_pick_suited_connectors = 12 * 4
'''
12개의 쌍 각각에 대해 4가지 무늬 조합이 있으므로, 12 × 4 = 48가지 조합
'''
# Probabilities
probability_pocket_pair = ways_to_pick_pocket_pair_per_rank / total_two_card_combinations
probability_connectors = ways_to_pick_connectors / total_two_card_combinations
probability_suited = ways_to_pick_suited / total_two_card_combinations
probability_suited_connectors = ways_to_pick_suited_connectors / total_two_card_combinations

print(probability_pocket_pair * 100)
print(probability_connectors * 100)
print(probability_suited * 100)
print(probability_suited_connectors * 100)


5.88235294117647
14.479638009049776
23.52941176470588
3.619909502262444


텍사스 홀덤에서 프리플랍 단계에서  
포켓 페어 : 확률은 약 5.88%입니다. 이는 두 장의 카드 중 같은 랭크를 받을 확률을 나타냅니다.  
커넥터 (Connectors): 확률은 약 14.48%입니다. 이는 두 장의 카드가 숫자가 연속적일 확률입니다.   
Suited (같은 무늬): 확률은 약 23.53%입니다. 이는 두 장의 카드가 같은 무늬일 확률입니다.  
suited connector (커넥터이면서 같은 무늬) : 확률은 약 3.62%입니다

In [3]:
from math import comb
# Calculating probabilities for Broadway Cards, 1-gap Connectors, 2-gap Connectors, Ax Hands, and Low Pairs pre-flop

# Total number of ways to pick any two cards from the deck
total_two_card_combinations = comb(52, 2)

# Broadway Cards (T, J, Q, K, A)
# There are 5 broadway ranks, each with 4 suits
ways_to_pick_broadway = comb(5 * 4, 2)
'''
T (10), J, Q, K, A를 포함하는 카드 조합입니다. 
이들은 특히 고순위 스트레이트를 형성할 수 있는 가능성 때문에 중요하며, 높은 원카드 가치를 지닙니다.
'''
# 1-gap Connectors (e.g., 5-7, 9-J)
# There are 11 possible 1-gap connectors (A not included with 3)
ways_to_pick_1_gap = 11 * 4 * 4  # Each can be of any suit

# 2-gap Connectors (e.g., 5-8, 9-Q)
# There are 10 possible 2-gap connectors (A not included with 4)
ways_to_pick_2_gap = 10 * 4 * 4  # Each can be of any suit
'''
Gap Connectors: 이는 숫자가 한 칸 이상 떨어진 커넥터를 말합니다. 
예를 들어, 5-7 또는 8-10 같은 카드입니다. 이들은 스트레이트를 만들 가능성은 낮지만 가능성이 여전히 있습니다. 
1-gap, 2-gap 등으로 구분할 수 있으며, 갭(gap)의 크기에 따라 확률이 달라집니다.
'''
# Ax Hands (Ace with any other card)
# There are 12 non-Ace ranks, each with 4 suits
ways_to_pick_ax = 4 * 12 * 4  # Ace with any other card of any suit
'''
Ace와 다른 어떤 카드의 조합입니다. 
에이스를 포함하고 있기 때문에 강력한 하이 카드를 가지고 있으며 
에이스가 포함된 핸드는 많은 상황에서 유용할 수 있습니다.
'''
# Low Pairs (2 through 6)
# There are 5 low ranks, and for each, there are 4 choose 2 ways to pick pairs
ways_to_pick_low_pairs = 5 * comb(4, 2)
'''
이는 낮은 숫자의 포켓 페어를 의미합니다 (예: 2-2, 3-3). 
이러한 페어는 큰 페어보다 확률적으로 약할 수 있지만, 세트를 만들어 게임을 뒤집을 수 있는 잠재력을 가지고 있습니다.
'''
# Probabilities
probability_broadway = ways_to_pick_broadway / total_two_card_combinations
probability_1_gap = ways_to_pick_1_gap / total_two_card_combinations
probability_2_gap = ways_to_pick_2_gap / total_two_card_combinations
probability_ax = ways_to_pick_ax / total_two_card_combinations
probability_low_pairs = ways_to_pick_low_pairs / total_two_card_combinations

print(probability_broadway * 100)
print(probability_1_gap * 100)
print(probability_2_gap * 100)
print(probability_ax * 100)
print(probability_low_pairs * 100)


14.328808446455504
13.273001508295627
12.066365007541478
14.479638009049776
2.262443438914027


### 플롭

In [4]:
from math import comb

# Assuming player starts with two suited cards
# There are 11 remaining cards of the same suit in the deck (13 total - 2 in hand)

# Number of ways to draw 3 cards of the same suit from the remaining 11 suited cards
ways_to_draw_flush = comb(11, 3)

# Total ways to draw any 3 cards from the remaining 50 cards (after excluding the 2 in hand)
total_ways_to_draw_flop = comb(50, 3)

# Probability of drawing a flush on the flop with two suited cards in hand
probability_flush_flop = ways_to_draw_flush / total_ways_to_draw_flop
'''
두 장의 suited 카드를 가지고 시작했을 때, 
플롭에서 플러시를 만들 확률은 약 0.84%입니다. 
이는 비교적 낮은 확률이지만, 이런 경우가 발생하면 매우 강력한 핸드가 됩니다.
'''

# Correct the calculation for the probability of making a set on the flop
# Assuming the player starts with a pocket pair and looking to hit a set

# Ways to draw one more card of the same rank from the remaining two of that rank
ways_to_draw_one_more_of_same_rank = comb(2, 1)

# Ways to draw the remaining two cards from the other 49 cards in the deck
ways_to_draw_two_other_cards = comb(49, 2)

# Total ways to draw any 3 cards from the remaining 50 cards (after excluding the pocket pair)
total_ways_to_draw_any_three_cards = comb(50, 3)

# Probability of making a set
probability_of_set_flop = (ways_to_draw_one_more_of_same_rank * ways_to_draw_two_other_cards) / total_ways_to_draw_any_three_cards
'''
플롭 단계에서 포켓 페어를 가지고 시작했을 때 세트(트리플)를 만들 확률은 약 12%입니다. 
이 확률은 플레이어가 포켓 페어를 가진 상태에서 플롭에 세 번째 같은 숫자의 카드를 받을 가능성을 나타냅니다. 
'''


print(probability_flush_flop * 100)
print(probability_of_set_flop * 100)


0.8418367346938775
12.0


In [9]:
# Define all possible connectors including Ace as low
connectors = [(v, v+1) for v in range(1, 14)]  # Now includes Ace (1, 2)

# Function to calculate straight sequences from a given connector
def calculate_straight_sequences(connector):
    sequences = []
    start, end = connector
    # Generate sequences that can form a straight containing both connector cards
    # Handle normal sequences and wrap-around for Ace as both high and low
    if start == 1:  # When Ace is low and connects with 2
        potential_sequences = [
            [1, 2, 3, 4, 5],  # A-2-3-4-5
        ]
    elif start == 14:  # When Ace is high and connects with 2
        potential_sequences = [
            [10, 11, 12, 13, 14],  # 10-J-Q-K-A
        ]
    else:
        potential_sequences = [
            [start - 3, start - 2, start - 1, start, end] if start > 3 else [],
            [start - 2, start - 1, start, end, end + 1] if start > 2 else [],
            [start - 1, start, end, end + 1, end + 2] if start > 1 else [],
            [start, end, end + 1, end + 2, end + 3],
        ]
    # Filter and validate sequences (all elements must be within the valid range and unique)
    for seq in potential_sequences:
        if seq and all(1 <= x <= 14 for x in seq) and len(set(seq)) == len(seq):
            sequences.append(seq)
    return sequences

# Calculate possible straight sequences for each connector
connector_straight_sequences = {connector: calculate_straight_sequences(connector) for connector in connectors}

# Calculate total combinations for each sequence to appear in the flop
total_ways_to_draw_straight_sequences = 0
for sequences in connector_straight_sequences.values():
    for sequence in sequences:
        # 4 cards for each rank, we need exactly these 3 cards from the sequence in the flop
        total_ways_to_draw_straight_sequences += 4**3  # Each card of the sequence can be any of 4 suits

# Total ways to draw any 3 cards from the remaining 50 cards (excluding the 2 held)
total_flop_combinations = comb(50, 3)

# Probability of making a straight on the flop for any connectors
probability_straight_flop_for_all_connectors_including_ace_low = total_ways_to_draw_straight_sequences / total_flop_combinations

probability_straight_flop_for_all_connectors_including_ace_low * 100
'''
1-gap 일때 플롭이 스트레이트일 확률
'''


13.061224489795919

In [8]:
connector_straight_sequences

{(1, 2): [[1, 2, 3, 4, 5]],
 (2, 3): [[1, 2, 3, 4, 5], [2, 3, 4, 5, 6]],
 (3, 4): [[1, 2, 3, 4, 5], [2, 3, 4, 5, 6], [3, 4, 5, 6, 7]],
 (4, 5): [[1, 2, 3, 4, 5], [2, 3, 4, 5, 6], [3, 4, 5, 6, 7], [4, 5, 6, 7, 8]],
 (5, 6): [[2, 3, 4, 5, 6], [3, 4, 5, 6, 7], [4, 5, 6, 7, 8], [5, 6, 7, 8, 9]],
 (6, 7): [[3, 4, 5, 6, 7], [4, 5, 6, 7, 8], [5, 6, 7, 8, 9], [6, 7, 8, 9, 10]],
 (7, 8): [[4, 5, 6, 7, 8],
  [5, 6, 7, 8, 9],
  [6, 7, 8, 9, 10],
  [7, 8, 9, 10, 11]],
 (8, 9): [[5, 6, 7, 8, 9],
  [6, 7, 8, 9, 10],
  [7, 8, 9, 10, 11],
  [8, 9, 10, 11, 12]],
 (9, 10): [[6, 7, 8, 9, 10],
  [7, 8, 9, 10, 11],
  [8, 9, 10, 11, 12],
  [9, 10, 11, 12, 13]],
 (10, 11): [[7, 8, 9, 10, 11],
  [8, 9, 10, 11, 12],
  [9, 10, 11, 12, 13],
  [10, 11, 12, 13, 14]],
 (11, 12): [[8, 9, 10, 11, 12], [9, 10, 11, 12, 13], [10, 11, 12, 13, 14]],
 (12, 13): [[9, 10, 11, 12, 13], [10, 11, 12, 13, 14]],
 (13, 14): [[10, 11, 12, 13, 14]]}

In [6]:
# Calculate the probability of making a straight on the flop when holding connectors, considering all sequences

# For connectors 6-7, possible straight sequences that complete the straight:
# These are the sequences: (3-4-5), (4-5-8), (5-8-9), (8-9-10)
# Each number has 4 suits, thus each sequence has 4^3 combinations

# Calculate combinations for each sequence
ways_to_draw_3_4_5 = 4**3
ways_to_draw_4_5_8 = 4**3
ways_to_draw_5_8_9 = 4**3
ways_to_draw_8_9_10 = 4**3

# Total ways to draw these sequences
total_ways_to_draw_straight_sequences = (
    ways_to_draw_3_4_5 + 
    ways_to_draw_4_5_8 + 
    ways_to_draw_5_8_9 + 
    ways_to_draw_8_9_10
)

# Total ways to draw any 3 cards from the remaining 50 cards (excluding the 2 held)
total_flop_combinations = comb(50, 3)

# Probability of making a straight on the flop
probability_straight_flop_all_cases = total_ways_to_draw_straight_sequences / total_flop_combinations

print(probability_straight_flop_all_cases * 100)


1.306122448979592
